When checking out a remote branch by name, ask the user how

stefanhaller opened the pull request:
- **PR Description**

When checking out a remote branch by name, ask the user how they want to check it out; the choices are to create a new local branch that tracks the remote, or a detached head.

This is an alternative to #3371, and fixes #2312.

- **Please check if the PR fulfills these requirements**

* [x] Cheatsheets are up-to-date (run `go generate ./...`)
* [x] Code has been formatted (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#code-formatting))
* [ ] Tests have been added/updated (see [here](https://github.com/jesseduffield/lazygit/blob/master/pkg/integration/README.md) for the integration test guide)
* [x] Text is internationalised (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#internationalisation))
* [ ] Docs (specifically `docs/Config.md`) have been updated if necessary
* [x] You've read through your own file changes for silly mistakes etc

stefanhaller commented:
We now have an inconsistency between the checkout-by-name command and pressing space on a remote branch. In the latter case it's not possible to checkout a detached head (I actually often want to do that, and the workaround is to hit enter on the branch and press space on the first commit, but that's a bit cumbersome). Instead, we get this prompt where you can type a name for the local branch; I have never typed anything in that prompt, I always hit enter to accept the default choice. So I wonder if we should replace that prompt with the same menu that I added here for checkout by name.

jesseduffield commented on pkg/i18n/english.go:
```diff
@@ -1064,6 +1067,9 @@ func EnglishTranslationSet() TranslationSet {
 		CheckoutByName:                      "Checkout by name",
 		CheckoutByNameTooltip:               "Checkout by name. In the input box you can enter '-' to switch to the last branch.",
+		RemoteBranchCheckoutType:            "Checkout type",
+		CheckoutTypeNewBranch:               "new local branch",
```
```suggestion
		CheckoutTypeNewBranch:               "New local branch",
```

jesseduffield commented on pkg/i18n/english.go:
```diff
@@ -1064,6 +1067,9 @@ func EnglishTranslationSet() TranslationSet {
 		CheckoutByNameTooltip:               "Checkout by name. In the input box you can enter '-' to switch to the last branch.",
+		RemoteBranchCheckoutType:            "Checkout type",
+		CheckoutTypeNewBranch:               "new local branch",
+		CheckoutTypeDetachedHead:            "detached head",
```
```suggestion
		CheckoutTypeDetachedHead:            "Detached head",
```

jesseduffield reviewed (approved):
Great work! Just a couple comments on casing, but otherwise LGTM

stefanhaller commented:
Thanks for the quick review, fixed the casing in feae89b07eb.

You didn't respond to the question about pressing space on a remote branch in the remote branches view. I thought about this more, and really want to change the behavior there to be consistent, especially because the old behavior is still available with the 'n' key. I added two commits to do this: a21db7a6b00 and 2e5b326bbd9.

And finally, I added a commit to show an inline waiting status when checking out a local branch (b20163cf5c7). I find that this improves the UX a bit, because I tend to have my eyes on the branches view so I easily miss the bottom-line status. To make this work well for checking out a new local branch we need to add an extra refresh so that the branch is visible (see df2accd05d7), but I think it's worth it.

Let me know your thoughts on all this!

jesseduffield commented:
> You didn't respond to the question about pressing space on a remote branch in the remote branches view. I thought about this more, and really want to change the behavior there to be consistent, especially because the old behavior is still available with the 'n' key. I added two commits to do this: https://github.com/jesseduffield/lazygit/commit/a21db7a6b0083a9edc33acf45a3e77e98e653a72 and https://github.com/jesseduffield/lazygit/commit/2e5b326bbd9cb3d0756b544a0f2a3c0ce9a37101.

My bad. Yep, let's change the behaviour there to make it consistent.

> And finally, I added a commit to show an inline waiting status when checking out a local branch (https://github.com/jesseduffield/lazygit/commit/b20163cf5c791f03aeb7701430f855e48992d8eb). I find that this improves the UX a bit, because I tend to have my eyes on the branches view so I easily miss the bottom-line status. To make this work well for checking out a new local branch we need to add an extra refresh so that the branch is visible (see https://github.com/jesseduffield/lazygit/commit/df2accd05d724e248f7855f39bcd416405dc78d1), but I think it's worth it.

That's awesome, I'm happy with that.

I've just pushed another commit adding some tooltips to the menu options and tweaking the title. I like the menu title to be based on what you're trying to do e.g. 'checkout X' so having it as 'checkout type' felt wrong. We really need a way to add a message inside the popup before the menu options e.g:
title: 'Checkout remote branch'
content: 'How would you like to checkout this branch?'
options: 'New local branch', 'Detached head'
Alas, that can wait

stefanhaller commented:
> I've just pushed another commit adding some tooltips to the menu options and tweaking the title.

Thanks for that. I found the tooltip about the detached head misleading (or wrong, actually; it's totally possible to create commits on a detached head, and it can sometimes be useful), so I changed it again in d9fdb89687e4. Hope you're ok with that text.

> We really need a way to add a message inside the popup before the menu options

Yes, this would be nice. I agree that it's not highest priority right now.
